public/app/features/dashboard/components/PanelEditor/state/actions.ts

```diff
@@ -56,9 +57,19 @@ export function exitPanelEditor(): ThunkResult<void> {
 
     const modifiedPanel = getPanel();
     const modifiedSaveModel = modifiedPanel.getSaveModel();
-    const initialSaveModel = getSourcePanel().getSaveModel();
-    const panelChanged = !isEqual(omit(initialSaveModel, 'id'), omit(modifiedSaveModel, 'id'));
-    if (shouldDiscardChanges || !modifiedPanel.libraryPanel || !panelChanged) {
+
+    if (shouldDiscardChanges || !modifiedPanel.libraryPanel) {
+      onConfirm();
+      return;
+    }
+
+    const libraryPanel = await getLibraryPanel(modifiedPanel.libraryPanel!.uid!);
+    const panelChanged = !isEqual(
+      omit(libraryPanel.model, 'id', 'libraryPanel', 'gridPos'),
+      omit(modifiedSaveModel, 'id', 'libraryPanel', 'gridPos')
+    );
```

Comment:
would like to move away from object diffing and move to a more deliberate dirty state flag set in PanelModel.updateOptions, updateFieldConfig, setTransforms, updateQueries (https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/components/PanelEditor/PanelEditorQueries.tsx#L61) 

Might be some that is not covered there, like the panel title, links & repeat options. Any ideas for how we can do that? 

@hugohaggmark what do you think? 

I want the same for DashboardModel as well (not related to this PR / Library panel), but more just some pain points we have with object diffing where properties are changed due to migrations and other non user inititated actions leading to dirty state the user as not actually performed.

Reply:
Ok, so if I understand correctly, we want a Redux mechanism to check if a Panel/Dashboard has been changed? Or can we use events?

For Redux I would create a middleware for that. The middleware listens for certain actions and changes state accordingly. The challenge is how that hooks into mutable parts; I guess? 

For the mutable parts I really only can come up with these: using Events, Proxy or get/set props to wrap PanelModel/DashboardModel.

Reply:
Now that I think about it, I wonder if having exitPanelEditor as a middleware instead would solve the `cancel the user flow` challenge we've been discussing?

Reply:
We cannot use redux easily without doing a really big refactor for PanelModel. I was just thinkgin a dirty flag on PanelModel that we set in PanelModel.updateOptions, PanelModel.setTransforms PanelModel.updateFieldConfig etc

Reply:
Aha sorry, I thought updateOptions, setTransforms and updateFieldConfig where Redux actions.